Migrate `validate_json.py` script to rust in `run-make/rustdoc-map-file` test

GuillaumeGomez opened the pull request:
This PR fixes the FIXME I added for future-me who become present-me. :')

Since there are multiple `run-make` tests using python scripts, I suppose more of them will migrate to Rust, hence why I added the `jzon` public reexport to the `run-make-support` crate.

cc @jieyouxu 
r? @Kobzol

GuillaumeGomez committed:
Migrate `validate_json.py` script to rust in `run-make/rustdoc-map-file` test

jieyouxu commented on Cargo.lock:
```diff
@@ -1939,6 +1939,12 @@ dependencies = [
  "serde_json",
 ]
 
+[[package]]
+name = "jzon"
+version = "0.12.5"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "17ab85f84ca42c5ec520e6f3c9966ba1fd62909ce260f8837e248857d2560509"
```
Question: is there a particular reason why we're using `jzon` or a new package, instead of something like `serde_json`? (I have not used `jzon` before)

Kobzol replied:
+1. Let's just use `serde_json::Value` for this to avoid adding a new dependency.

GuillaumeGomez replied:
Ah indeed. Let's go with `serde_json` then!

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
```diff
@@ -8,6 +10,48 @@ fn main() {
-    // FIXME (GuillaumeGomez): Port the python script to Rust as well.
-    python_command().arg("validate_json.py").arg(&out_dir).run();
+
+    let mut found_file = false;
+    read_dir_entries_recursive(&out_dir, |path| {
+        if !found_file
+            && path.is_file()
+            && path.file_name().map(|name| name == "redirect-map.json").unwrap_or(false)
```
Question: are there multiple `redirect-map.json`s generated? Or is the directory structure in which they are going to be generated as unstable / subject to frequent changes?

GuillaumeGomez replied:
I was simply lazy to look at the right place. ^^'

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
```diff
@@ -8,6 +10,48 @@ fn main() {
+        {
+            found_file = true;
+            let generated = read_to_string(path);
+            let expected = read_to_string("expected.json");
```
Question (following the recursive directory traversal): But AFAICT, there is only one reference file, so I'm a bit confused.

GuillaumeGomez replied:
We iterate through folders only to find the generated file. The expected one doesn't move.

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
```diff
@@ -8,6 +10,48 @@ fn main() {
+                match generated.get(key) {
+                    Some(value) => {
+                        if expected_value != value {
+                            differences.push(format!("values for key `{key}` don't match"));
```
Question: do we also want to provide info on *how* they don't match upon test failure?

GuillaumeGomez replied:
Good idea!

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
```diff
@@ -8,6 +10,48 @@ fn main() {
+            }
+            for (key, data) in generated.entries() {
+                if !expected.has_key(key) {
+                    differences
+                        .push(format!("Extra data not expected: key: `{key}`, data: `{data}`"));
```
Remark: ah but here we *do* provide info on what data is extra.

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
```diff
@@ -8,6 +10,48 @@ fn main() {
+                for diff in differences {
+                    eprintln!("=> {diff}");
+                }
+                std::process::exit(1);
```
Question: do we want to `process::exit` here, or do we want to `panic!`?

GuillaumeGomez replied:
I thought the `panic` was a bit too much but it might be better for cross-platform testing to use `panic`.

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
Note: I have absolutely no clue what the rustdoc map file does, so bear with me here :3

jieyouxu commented on tests/run-make/rustdoc-map-file/rmake.rs:
Suggestion: we should add a bit of context for what this test is trying to check, are there any further context (like issues/discussions/FCPs etc.)?

jieyouxu replied:
@rustbot author

jieyouxu reviewed (commented):
Cool! I have a few questions about `jzon` and the test itself, but I'm glad it's viable to get rid of the python script :3

jieyouxu commented:
r? jieyouxu

GuillaumeGomez committed:
Reexport `serde_json` crate from run-make-support to give it access to `run-make` tests

GuillaumeGomez commented:
Sum up of what I did:

* I replaced `jzon` with `serde_json` (still reexported).
* I stopped recursively going through folders and simply get the file from a known path (which is actually better since if we change the path, better to know it).
* Removed usage of `exit` and only use `panic` instead.
* Added a comment explaining what the test is checking.

GuillaumeGomez commented:
Ah I forgot the improvement for error messages. Done as well. :)

jieyouxu commented on src/tools/run-make-support/src/lib.rs:
```diff
@@ -7,6 +7,8 @@
 // skip rustfmt on this file.
 #![cfg_attr(rustfmt, rustfmt::skip)]
 
+pub use serde_json;
+
```
Suggestion: there is a smol section of pub re-exports of third-party libs just a few lines below, could we move it there?

https://github.com/rust-lang/rust/blob/7116ed4983119b6506d954c6bad64f07b3017c83/src/tools/run-make-support/src/lib.rs#L38

GuillaumeGomez replied:
Ah sure!

jieyouxu reviewed (approved):
Thanks, this looks good! Just one minor style nit on the re-export. Feel free to r=me after the nit once PR CI is green.

GuillaumeGomez commented:
Moved the reexport with the others.

jieyouxu commented on src/tools/run-make-support/src/lib.rs:
```diff
@@ -7,6 +7,8 @@
 // skip rustfmt on this file.
 #![cfg_attr(rustfmt, rustfmt::skip)]
 
+pub use serde_json;
+
```
Remark: is github diff outdated here or is this duplicated?

GuillaumeGomez replied:
It was actually simpler: I messed up my rebase. Fixed.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10420798540/job/28861564850) [(plain)](https://github.com/rust-lang/rust/commit/4cbd9cebf848f75fa05eb4ea9818dca3d59d4b06/checks/28861564850/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:20d3b4d4a2629cbf7865cdbf92fe47512a7c96658c24253a045ff38e8075cd7fb37ca6fcadfa6e6d093333943ad24f6fc4f163ec5b74fd940de9d5bb03eb4d3b:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.72s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
   |         ^^^^^^^^^^

error: could not compile `run_make_support` (lib) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Fri Aug 16 14:23:28 UTC 2024
  network time: Fri, 16 Aug 2024 14:23:29 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

jieyouxu commented:
... but that seems to work now, so... 🤷 

@bors r+ rollup

bors commented:
:pushpin: Commit 6b1637c477f9f40821c68a9452f7435d9707948d has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 6b1637c477f9f40821c68a9452f7435d9707948d -->
<!-- homu: {"type":"Approved","sha":"6b1637c477f9f40821c68a9452f7435d9707948d","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

Zalathar commented:
Heh, I was also working on a version of this, and only noticed when my own changes wouldn't rebase cleanly onto master. 😅

My approach was to add a `Diff::normalize_json` method that pretty-prints the "actual" text before diffing it, which lets us get rid of all the ad-hoc JSON-diffing code. I might still write up my own PR because I think it's a modest improvement over this version, and a similar technique can be used to get rid of the (trivial) Python script in `run-make/libtest-json`.

GuillaumeGomez commented:
I'm not sure it's worth it to have anything more advanced than this but don't let it stop you. 😉

tgross35 commented:
Bors is not correctly removing items from the queue

@bors r-
